gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/BaseDatastoreBatchWriter.java

```diff
@@ -64,10 +67,8 @@ public final void addWithDeferredIdAllocation(FullEntity<?>... entities) {
 
   private void addInternal(FullEntity<Key> entity) {
     Key key = entity.key();
-    if (toAdd.containsKey(key) || toUpdate.containsKey(key) || toPut.containsKey(key)) {
-      throw newInvalidRequest("Entity with the key %s was already added or updated in this %s",
-          entity.key(), name);
-    }
+    checkArgument(!toAdd.containsKey(key) && !toUpdate.containsKey(key) && !toPut.containsKey(key),
```

Comment:
This is actually changing the possible set of exceptions thrown by the Datastore.

Before it was DatastoreException with DatastoreError.FAILED_PRECONDITION as error code
and now IllegalArgumentExcetpion.

Though I think this change is reasonable (and aligned with other services) I am not sure it must
be done in this PR (I would rather have a separate issue for such change).
In any case, when we change that we should also change the "@throws" sets of the relevant methods.